Move operators to utilities

gassmoeller opened the pull request:
#1720 introduced some duplication by copying operators into both interfaces for initial composition and initial temperature. As I plan to extend this functionality to boundary conditions as well, it makes sense to move the code into utilities instead, and make it available for other parts of the program as well.

bangerth commented on include/aspect/utilities.h:
```diff
@@ -955,6 +956,63 @@ namespace aspect
+         * Execute the selected operation with the given parameters and
+         * return the result.
+         */
+        double operator() (double x, double y) const;
```
make these `const`.

bangerth commented on include/aspect/utilities.h:
```diff
@@ -955,6 +956,63 @@ namespace aspect
+        /**
+         * The selected operation of this object.
+         */
+        operation op;
```
could this be `const` as well?

bangerth replied:
or does the object need to be copyable?

gassmoeller replied:
If I make this `const` I need to provide at least another `operator =` because the default move assignment operator does not work anymore. I do not know what else will be needed, so I kept it non-const for now. I also thought that it might be nice to provide a way to switch the operation later (by adding a set- function), but I am not feeling strongly about it, it was just simpler to implement this way.

bangerth replied:
Not worth worrying about then.

If you do make it `const`, no `operator=` can change that, of course :-)

bangerth commented on source/initial_composition/interface.cc:
```diff
@@ -180,36 +169,7 @@ namespace aspect
-              }
-            }
-
+          composition = model_operators[i](composition,(*initial_composition_object)->initial_composition(position,  n_comp));
```
can you break the arguments onto two lines?

bangerth commented on source/initial_temperature/interface.cc:
```diff
@@ -171,36 +157,7 @@ namespace aspect
-              }
-            }
-
+          temperature = model_operators[i](temperature,(*initial_temperature_object)->initial_temperature(position));
```
same here; or put a space after the comma to make it easier to read

bangerth commented on source/utilities.cc:
```diff
@@ -2363,6 +2364,84 @@ namespace aspect
+          default:
+          {
+            Assert ( false, ExcMessage ("ASPECT's operator class only supports the following operators: "
+                                        "add, subtract, minimum and maximum. Please check your parameter file.") );
```
You shouldn't actually get here -- something must have gone wrong with parsing before. I would use `ExcInternalError` here.

gassmoeller replied:
I can get here if somebody used the default constructor, and I can not remove the default constructor, because I currently use it in `create_model_operator_list` to resize a std::vector. I could work around this, if you want to make sure this never happens.

bangerth replied:
But that's a programming error, not something that has to do with user input files. So I'd still favor `ExcInternalError()`.

bangerth commented on source/utilities.cc:
```diff
@@ -2363,6 +2364,84 @@ namespace aspect
+          case Utilities::Operator::add:
+          {
+            return x + y;
+            break;
```
no need for `break` after `return`. Same below.

gassmoeller commented:
Addressed all of the comments and fixed the compiler errors (I needed to include new headers in utilities.h now that simulator_access.h is not longer included).

bangerth commented on include/aspect/utilities.h:
```diff
@@ -955,6 +957,63 @@ namespace aspect
+          subtract,
+          minimum,
+          maximum,
+          uninitialized
```
I should have seen that in the first round: put `uninitialized` first in the list; this way, the compiler assigns the numeric value zero to it, and that's what the field will be default-initialized to -- which is what you want if one forgets an explicit initializer.

bangerth commented:
OK to merge if you make the one modification.
